invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/bech32"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/roasbeef/btcd/btcec"
+)
+
+const (
+	mSatPerBtc = uint64(100000000000)
+
+	// The byte corresponding to the supported field types. The field name
+	// is the character representing that 5-bit value in the bech32 string.
+	fieldTypeP = byte(1)
+	fieldTypeD = byte(13)
+	fieldTypeN = byte(19)
+	fieldTypeH = byte(23)
+	fieldTypeX = byte(6)
+	fieldTypeF = byte(9)
+	fieldTypeR = byte(3)
+)
+
+var (
+
+	// unitToMsat is a map from a unit to a function that converts an amount
+	// of that unit to millisatoshis.
+	unitToMsat = map[byte]func(uint64) (uint64, error){
+		'm': func(am uint64) (uint64, error) {
+			return am * 100000000, nil
+		},
+		'u': func(am uint64) (uint64, error) {
+			return am * 100000, nil
+		},
+		'n': func(am uint64) (uint64, error) {
+			return am * 100, nil
+		},
+		'p': func(am uint64) (uint64, error) {
+			if am < 10 {
+				return 0, fmt.Errorf("minimum amount is 10p")
+			}
+			if am%10 != 0 {
+				return 0, fmt.Errorf("amount not expressible " +
+					"in msat")
+			}
+			return am / 10, nil
+		},
+	}
+
+	// msatToUnit is a map from a unit to a function that converts an amount
+	// in millisatoshis to an amount of that unit.
+	msatToUnit = map[byte]func(uint64) (uint64, error){
+		'm': func(am uint64) (uint64, error) {
+			if am%100000000 != 0 {
+				return 0, fmt.Errorf("%d msat not expressible "+
+					"in mBTC", am)
+			}
+			return am / 100000000, nil
+		},
+		'u': func(am uint64) (uint64, error) {
+			if am%100000 != 0 {
+				return 0, fmt.Errorf("%d msat not expressible "+
+					"in uBTC", am)
+			}
+			return am / 100000, nil
+		},
+		'n': func(am uint64) (uint64, error) {
+			if am%100 != 0 {
+				return 0, fmt.Errorf("%d msat not expressible "+
+					"in nBTC", am)
+			}
+			return am / 100, nil
+		},
+		'p': func(am uint64) (uint64, error) {
+			return am * 10, nil
+		},
+	}
+)
+
+// MessageSigner is passed to the Encode method to provide a signature
+// corresponding to the node's pubkey.
+type MessageSigner struct {
+	// SignCompact signs the passed hash with the node's privkey.
+	SignCompact func(hash []byte) ([]byte, error)
+}
+
+// Invoice represents a decoded invoice, or to-be-encoded invoice. Some of the
+// fields are optional, and will only be non-nil if the invoice this was parsed
+// from contains that field. When encoding, only the non-nil fields will be
+// added to the encoded invoice.
+type Invoice struct {
+	// Net specifies what network this Lightning invoice is meant for.
+	Net *chaincfg.Params
+
+	// MilliSat specifies the amount of this invoice in milli-satoshi.
+	// Optional.
+	MilliSat *uint64
+
+	// Timestamp specifies the time this invoice was created.
+	Timestamp uint64 // Mandatory
+
+	// PaymentHash is the payment hash to be used for a payment to this
+	// invoice.
+	PaymentHash []byte
```

Comment:
Since an invoice must have a payment hash, we need to be able to after parsing check if the payment hash was found and set, so I ended up using the type `*[32]byte`. I could have compared the payment hash to an array filled with zeroes, but don't know if this is the proper way?